### PR TITLE
[16.0][MIG] sale_order_import: Migration to 16.0

### PR DESCRIPTION
**Pull request by QuocDuong1306**

### CODE REVIEW

File: sale_order_import/wizard/sale_order_import.py [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+
+class SaleOrderImport(models.TransientModel):
+    _name = "sale.order.import"
+    _description = "Sale Order Import from Files"
+
+    state = fields.Selection(
+        [("import", "Import"), ("update", "Update")], default="import"
+    )
+    partner_id = fields.Many2one("res.partner", string="Customer")
+    csv_import = fields.Boolean(default=False, readonly=True)
```
Comment (simahawk) [inline sale_order_import/wizard/sale_order_import.py]: This flag should be removed and a new selection field `import_type` should be added.
`supported_types` below should depend on this and a `_parse_{import_type}` should be used by `_parse_file`

File: sale_order_import/wizard/sale_order_import.py [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+        except etree.XMLSyntaxError:
+            error_msg = _("This XML file is not XML-compliant")
+            return xml_root, error_msg
+        try:
+            self.parse_xml_order(xml_root, detect_doc_type=True)
+        except (UserError, NotImplementedError):
+            error_msg = _("Unsupported XML document")
+        return xml_root, error_msg
+
+    # FIXME: not used at all
```
Comment (simahawk) [inline sale_order_import/wizard/sale_order_import.py]: FIXME pls :)

File: sale_order_import/wizard/sale_order_import.py [hunk 3]
```diff
@@ -0,0 +1,663 @@
...
+    @api.model
+    def parse_xml_order(self, xml_root, detect_doc_type=False):
+        raise NotImplementedError(
+            _(
+                "This type of XML RFQ/order is not supported. Did you install "
+                "the module to support this XML format?"
+            )
+        )
+
+    # FIXME: not used at all
```
Comment (simahawk) [inline sale_order_import/wizard/sale_order_import.py]: FIXME pls

File: sale_order_import/wizard/sale_order_import.py [hunk 4]
```diff
@@ -0,0 +1,663 @@
...
+    def parse_csv_order(self, order_file, partner):  # pragma: no cover
+        assert partner, "missing partner"
+        raise UserError(
+            _(
+                "This type of CSV order is not supported. Did you install "
+                "the module to support CSV orders?"
+            )
+        )
+
+    # TODO: move it out to a PDF support module
```
Comment (simahawk) [inline sale_order_import/wizard/sale_order_import.py]: this is a new import type  in theory. Yet, as is still based on xml parsing in the end, we can leave it here but refactored according to the "import_type" mentioned above

File: sale_order_import/wizard/sale_order_import.py [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+            {
+                "view_mode": "form,tree,calendar,graph",
+                "views": False,
+                "view_id": False,
+                "res_id": order.id,
+            }
+        )
+        return action
+
+    # TODO: add tests
```
Comment (simahawk) [inline sale_order_import/wizard/sale_order_import.py:425]: TODO

File: sale_order_import/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    def name_get(self):
+        """Add amount_untaxed in name_get of sale orders"""
+        res = super().name_get()
+        if self._context.get("sale_order_show_amount"):
+            new_res = []
+            for (sale_id, name) in res:
+                sale = self.browse(sale_id)
+                # I didn't find a python method to easily display
+                # a float + currency symbol (before or after)
```
Comment (simahawk) [inline sale_order_import/models/sale.py]: the ir.qweb.field.monetary widget should be used probably

**Review by simahawk — CHANGES_REQUESTED** [review]
if you add the changes that I proposed, pls make them in separated commits :pray:

File: sale_order_import/tests/test_parse_validate.py [hunk 7]
```diff
@@ -15,18 +15,17 @@
...
-        self.assertTrue(self.wiz_model._unsupported_file_msg("fname.omg"))
-        # Test it gets called (cannot do it w/ Form)
-        mock_file_msg = mock.patch.object(type(self.wiz_model), "_unsupported_file_msg")
-        with mock_file_msg as mocked:
+    def test_wrong_import_type(self):
+        order_file_data = base64.b64encode(
+            b"<?xml version='1.0' encoding='utf-8'?><root><foo>baz</foo></root>"
+        )
+        order_filename = "test_order.xml"
+        with self.assertRaises(exceptions.UserError):
```
Comment (simahawk) [inline sale_order_import/tests/test_parse_validate.py]: pls check for the specific error msg if possible

File: sale_order_import/wizard/sale_order_import.py [hunk 8]
```diff
@@ -72,18 +77,43 @@ def _parse_file(self, filename, filecontent, detect_doc_type=False):
         logger.debug("Order file mimetype: %s", filetype)
         mimetype = filetype[0]
         supported_types = {
```
Comment (simahawk) [inline sale_order_import/wizard/sale_order_import.py]: to ease plugging new import types, this should stay in a separated method :wink:

File: sale_order_import/wizard/sale_order_import.py [hunk 9]
```diff
@@ -72,18 +77,43 @@ def _parse_file(self, filename, filecontent, detect_doc_type=False):
...
+            raise UserError(
+                _(
+                    "This file '{filename}' is not recognized as a {type} file. "
+                    "Please check the file and its extension.",
+                    filename=filename,
+                    type=self.import_type.upper(),
+                )
+            )
+        # Call the appropriate parsing method based on the selected import type
+        if self.import_type == "xml":
```
Comment (simahawk) [inline sale_order_import/wizard/sale_order_import.py]: IMO these checks are not needed. If yes, do them in their specific parsers

File: sale_order_import/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -0,0 +1,7 @@
+This module adds support for the import of electronic RFQ or orders. This module provides the base methods to import electronic orders ; it requires additional modules to support specific order formats:
+
+
+* module *sale_order_import_ubl*: adds support for `Universal Business Language (UBL) <http://ubl.xml.org/>`_ RFQs and orders as:
+
+  - XML file,
+  - PDF file with an embedded XML file.
```
Comment (simahawk) [inline sale_order_import/readme/DESCRIPTION.rst:7]: mention that you can also plug additional formats by extending the wizard :wink:

**Review by simahawk — CHANGES_REQUESTED** [review]
Thanks for the improvements! Few remarks...

File: sale_order_import/models/sale.py (hunk 6 above)
Comment (simahawk) [reply sale_order_import/models/sale.py]: can you add at least a TODO for my comment above :arrow_up:

File: sale_order_import/tests/test_parse_validate.py (hunk 7 above)
Comment (simahawk) [reply sale_order_import/tests/test_parse_validate.py]: ?

**Review by simahawk — APPROVED** [review]
LG overall

**Comment by QuocDuong1306** [comment]
@simahawk , I updated

**Review by simahawk — APPROVED** [review]
LG

**Review by yvaucher — APPROVED** [review]
LGTM

**Comment by yvaucher** [comment]
Note: we can search for a better method for the TODO about currency later, I'm not aware of one neither.

**Pull request merged**